File: 2ndPeriod/aedsII/tps/tp01/11/PalindromeRecursive.c
/* 
 * Nome: Guilherme Soares Silva
 * Matrícula: 863485
 * Disciplina: Algoritmos e Estruturas de Dados II
 * Atividade: C TP01Q02 - Palindromo
 * Data: 30/08/2024
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <ctype.h>
#include <stdbool.h>

#define MAX_LIMIT 256

bool isPalindromeRecursive(char *, int, int);
void containsSpecialChars(char *);

int main(void) {
    char str[MAX_LIMIT];
    bool loop = true;

    while(loop) {
        fgets(str, sizeof(str), stdin);
        str[strcspn(str, "\n")] = '\0'; 
        if(strcmp(str, "FIM") == 0) {
            loop = false;
        } else {
            containsSpecialChars(str);
            printf("%s\n", isPalindromeRecursive(str, 0, strlen(str) - 1) ? "SIM" : "NAO");
        }
    }
    return 0;
}

// Função recursiva para verificar se algum caractere especial está presente na string
void containsSpecialChars(char *s) {
    if (*s == '\0') {
        return;
    }
    if (!(((*s >= 'A') && (*s <= 'Z')) || ((*s >= 'a') && (*s <= 'z')))) {
        if ((*s >= '0') && (*s <= '9')) {
            *s = *s - 1;
        } else {
            *s = 'z';
        }
    }
    containsSpecialChars(s + 1);
}

// Função recursiva para verificar se a palavra é um palíndromo
bool isPalindromeRecursive(char *s, int left, int right) {
    if (left >= right) {
        return true;
    }
    if (s[left] != s[right]) {
        return false;
    }
    return isPalindromeRecursive(s, left + 1, right - 1);
}

File: 2ndPeriod/aedsII/tps/tp01/10/PalindromeRecursive.java
/* 
 * Nome: Guilherme Soares Silva
 * Matrícula: 863485
 * Disciplina: Algoritmos e Estruturas de Dados II
 * Atividade: JAVA TP01Q10 - RECURSIVO - Palíndromo
 * Data: 25/08/2024
 */
import java.util.Scanner;

class PalindromeRecursive {
    // Verifica se a palavra é 'FIM'
    public static boolean isEnd(String s) {
        return (s.length() == 3 && s.charAt(0) == 'F' && s.charAt(1) == 'I' && s.charAt(2) == 'M');
    }

    // Chama a função verificadora com a posição inicial e final da string
    public static boolean isPalindrome(String s) {
        int i = 0, f = s.length() - 1;
        return isPalindrome(i, f, s);
    }
    // Verifica se a palavra é um palíndromo
    public static boolean isPalindrome(int i, int f, String s) {
        if (i >= f) {
            return true;
        }
        if (s.charAt(i) != s.charAt(f)) {
            return false;
        }
        return isPalindrome(i + 1, f - 1, s);
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String str, result;
        boolean loop = true;

        // Loop para a leitura e print das palavras
        while (loop) {
            str = sc.nextLine();
            if (isEnd(str)) {
                loop = false;
            } else {
                result = isPalindrome(str) ? "SIM" : "NAO";
                System.out.println(result);
            }
        }
        sc.close();
    }
}